[Merged by Bors] - feat(algebra/big_operators/intervals): summation by parts

ecstatic-morse opened the pull request:
Add the "summation by parts" identity over intervals of natural numbers, as well as some helper lemmas.

---

It's generally useful (I want it for Dirichlet's convergence test) but I couldn't find it anywhere. Hopefully I didn't miss it; that `calc` proof took a while to write :smile:. I've split the file into sections since the existing variable declarations weren't quite right and I needed some `local notation` to make things more manageable.

I haven't touched any of the existing lemmas for now. Some of them are also restricted to intervals of natural numbers, so presumably they could be moved next to the stuff from this PR. Alternatively I could generalize everything, but I don't have an intuitive understanding of the typeclass hierarchy yet. Let me know how it can be improved. This is only my second non-trivial PR.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ecstatic-morse committed:
feat(algebra/big_operators/intervals): summation by parts

Add the summation by parts identity for intervals of natural numbers.

b-mehta commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +166,119 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+begin
+  convert sum_empty,
+  exact Ico_self n,
+end
```
It looks like this lemma and the one below can be proved by simp, or just a sequence of rewrites, so perhaps they don't need to be their own lemmas?

ecstatic-morse replied:
I need to rewrite `f n` to `sum (Ico m (m+1)) f` when proving another lemma, hence why `sum_Ico_single` exists. I think I stuck this one in for completion's sake. If I were to remove both of them, I would have to inline `sum_Ico_single` in the lemma that depends on it, or find another way to write the proof. Is this what you want?

ecstatic-morse replied:
These have been removed.

b-mehta commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +166,119 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+by rw [sum_eq_sum_Ico_succ_bot (lt_add_one n), sum_Ico_empty, add_zero]
+
+lemma sum_Ico_sub {c: ℕ} (hm : c ≤ m) (hmn : m ≤ n) :
+  ∑ (i : ℕ) in Ico (m-c) (n-c), f (i+c) = ∑ (i : ℕ) in Ico m n, f i :=
```
```suggestion
  ∑ (i : ℕ) in Ico (m-c) (n-c), f (c+i) = ∑ (i : ℕ) in Ico m n, f i :=
```
This saves you the add_comm rewrite, and brings the lemma in line with `sum_Ico_add`. Also I think you can move this one up to live next to that?

ecstatic-morse replied:
Why is it desirable to put `c` on the right in the index but on the left in the summand? I assumed this was just an oversight from the author of `sum_Ico_add` (perhaps their motivating use-case happened to be in this form?). Is there some logic behind this?

ecstatic-morse replied:
@b-mehta. There are now two versions of `sum_Ico_add`, one with the current ordering of `c` (to avoid breakage) and one where `c` appears consistently on the right. I also simplified the proofs. I think these lemmas are poorly named (this is not the most important result involving summation and addition), but it's consistent I guess.

I removed `sum_Ico_sub` for now, since there seems to be a strong preference for minimal lines of code/supporting lemmas among Mathlib reviewers. I'm not sure that's the correct choice, since it's more difficult to do rewrites and library search. However, my primary goal is to get this PR merged so I can start upstreaming the stuff that depends on it.

b-mehta commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +166,119 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+section group
+variable [add_comm_group β]
+
+lemma sum_Ico_succ_sub_sum_eq_top : F (n+1) - F n = f n :=
```
It looks like the multiplicative version of this lemma ought to be true too (ie `F (n+1) / F n = f n` but with `F` as a product), can you prove that and then use @[to_additive] to derive this version? Similarly for the other lemmas in this section

ecstatic-morse replied:
I tried this briefly this morning, but of course the product version is not true when `F n` is `0` (likewise for `f n` in the one below). How does `to_additive` handle this? Do I have to write two separate versions?

ecstatic-morse replied:
This question is still relevant: do you want me to write separate multiplicative versions with `(h : F n != 0)`? I don't need them in this PR.

j-loreaux replied:
This is what he means. The point is the multiplicative versions hold in a `comm_group` (instead of an `add_comm_group`), but they don't hold in a `ring` (or a `comm_group_with_zero`). But doing it this way gets you two lemmas for the price of one.
```lean
section group

-- The partial products (starting from `0`) of `f` and `g` respectively.
local notation `F` n:80 := ∏ i in range n, f i
local notation `G` n:80 := ∏ i in range n, g i

variable [comm_group β]

@[to_additive]
lemma prod_range_succ_div_prod : F (n+1) / F n = f n :=
div_eq_iff_eq_mul'.mpr $ prod_range_succ f n

@[to_additive]
lemma prod_range_succ_div_top : F (n+1) / f n = F n :=
div_eq_iff_eq_mul.mpr $ prod_range_succ f n

@[to_additive]
lemma prod_Ico_div_bot (hmn : m < n) : (∏ i in Ico m n, f i) / f m = ∏ i in Ico (m+1) n, f i :=
div_eq_iff_eq_mul'.mpr $ prod_eq_prod_Ico_succ_bot hmn _

@[to_additive]
lemma prod_Ico_succ_div_top (hmn : m ≤ n) :
  (∏ i in Ico m (n+1), f i) / f n = ∏ i in Ico m n, f i :=
div_eq_iff_eq_mul.mpr $ prod_Ico_succ_top hmn _

end group
```
Note, if you do `#check sum_Ico_succ_sub_top` after this, you will see the additive version. Also, note that you were missing some parentheses in the latter two lemmas.

b-mehta commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +166,119 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+
+/-- **Summation by parts**, also known as **Abel's lemma** or an **Abel transformation** -/
+theorem sum_by_parts (hmn : m < n) :
+  ∑ i in Ico m n, (f i * g i) =
```
```suggestion
  ∑ i in Ico m n, f i * g i =
```
The statement of this lemma looks sensible to me, but I wonder if the proof can be simplified!

ecstatic-morse replied:
Perhaps you can help me understand the culture of proof golfing that seems to be the standard in Mathlib?

I totally understand minimizing the number of "helper" lemmas in the global namespace that aren't broadly useful (like `sum_Ico_single` above, although it is useful, just duplicative with `sum_singleton`), and there are situations where a better technique can dramatically shorten a proof ([like the use of asymptotics in this thread](https://leanprover.zulipchat.com/login/#narrow/stream/113488-general/topic/.E2.9C.94.20Proving.20.60tendsto.60.20lemmas.20with.20filters)). However, I don't see the point of rewriting things just to minimize the number of lines.

The proof of this identity is fundamentally a series of tedious algebraic manipulations; I doubt there's a technique in Lean that would change that fact. I could shorten it by replacing the `calc` proof with a combination of `simp`, `rewrite`, `conv` and `nth_rewrite`-a Lean expert could certainly do even better-but the whole point of `calc` is to write proofs that are longer but easier to repair since each step is recorded explicitly. Admittedly this particular one is kind of unreadable due to the 100 character line limit. I don't find the argument about compilation-time compelling. This `calc` proof compiles pretty fast (it's just rewrites and `ring` after all). Meanwhile there are [hundreds of unexpanded invocations of `tidy`](https://leanprover.zulipchat.com/login/#narrow/stream/113488-general/topic/.60tidy.60.20in.20mathlib.20proofs) in Mathlib to finish proofs, some of which take several seconds apiece.

To be clear, these questions don't arise solely from your comment. I've written several Mathlib PRs over the last few weeks that have received suggestions for shortening proofs by combining rewrites, eliminating variables, or switching to term mode. Is golfing proofs skill that I'll need to develop if I want to continue contributing to Mathlib? Should I avoid `calc` in general, since it tends to result in longer proofs? Am I ignoring some constraints that makes these kind of transformations desirable?

j-loreaux replied:
I think the culture of golfing probably has a lot of sources. However, there are plenty of benefits. First, by trying to golf a proof, you gain a better understanding of what the key steps / transformations are, and which are not so important. Sometimes you even find that steps you thought were important actually weren't, and you can even go from a tactic mode proof to a one-line term proof.

Second, mathlib isn't a textbook for learning, it's more like a reference tome. It contains generic results and the proofs are written for experts, not students, so we don't need to show steps. Therefore, showing each step, _especially_ tedious steps is undesirable, so chaining together a bunch of rewrites into a single list is almost always nice. This often makes code more readable, not less.

Third, golfing can be fun. I'm sure there are plenty of other reasons besides these.

As for `tidy` or `simp` versus `simp only`, there is a trade-off between compilation time and simplicity/beauty/comprehensibility of the resulting code. This is likely one of the reasons for lots of `tidy` invocations.

ecstatic-morse replied:
@j-loreaux, all those points make sense in the broader context of Mathlib. The search for the most "elegant" proof has been a part of mathematics long before proof assistants were. I guess I'm trying to understand what @b-mehta is asking for specifically when they say:

>  I wonder if the proof can be simplified!

Of course it can! If term-mode proofs are considered "simpler" than tactic mode proofs, then *every* tactic mode proof can be simplified.

Your second point is about aesthetics (readibility/simplicity/beauty/comprehensibility). Does this proof violate some aesthetic principles that I'm not aware of? Should I collapse a few `calc` lines to avoid showing tedious steps? Switch away from `calc` entirely? Is there a tactic that could be used here that I don't know about? I don't want to spend time simplifying something only to be told that I'm doing it the wrong way.

In your first point, you mention that 

> by trying to golf a proof, you gain a better understanding of what the key steps / transformations are

but I don't want to get a better understanding of how to prove summation by parts, and doing so certainly won't yield a one-line term proof. Summation by parts is boring. I just want it to exist in the library so I can prove something slightly less boring.

> mathlib isn't a textbook for learning

I wrote the proof this way because I thought the fastest way for me to write it was to copy-paste each step, change it incrementally to match what I had on paper, and then prove that the incremental change was correct. I didn't intend for it to be useful for learning. If this way of doing things is not suitable for inclusion in Mathlib, then I will change how I do things. By understanding why golfing is important, I hope to understand what kind of proofs *are* suitable, and to write those instead.

j-loreaux replied:
First, please note that I am not criticizing you, only providing information. I can't speak for @b-mehta, or even more generally for mathlib maintainers, but I far as I am concerned and according to my own experience:

> If term-mode proofs are considered "simpler" than tactic mode proofs

Not necessarily. A 3-line term proof that could be closed in tactic mode via `by simp` is not preferable to the `simp` proof. On the other hand, a long calc proof which is boring for humans (but shows all the type transformations) may not be preferable to a term proof.

> Should I collapse a few `calc` lines to avoid showing tedious steps?

Definitely, especially successive rewrites, but also maybe find ways to restructure so that you don't have to do so many calculations "by hand".

> Switch away from `calc` entirely?

Not necessarily here, and definitely don't always avoid it, but use whatever works best, which is obviously context dependent.

> I wrote the proof this way because I thought the fastest way for me to write it was to copy-paste each step, change it incrementally to match what I had on paper, and then prove that the incremental change was correct.

That can be a reasonable way to write a first proof, but then afterwards you want to clean it up as much as possible. By this I mean golfing, or maybe writing a few key steps with `have`s, or looking for bits in the library that you can apply. Actually, this last bit is one reason _not_ to go straight from paper to Lean, because the existing API may affect decisions regarding proof structure. Of course, in this particular case the API may be limited, I'm not sure.

I could be wrong, but I think Bhavik's comment may be related to my first reaction to seeing the proof: "wow, that seems long. There's probably a simpler and shorter way." But without delving into the details of the proof and trying to rewrite some of it myself, I'm not immediately sure where to simplify (except for trivial things like chaining rewrites).

ecstatic-morse replied:
I am still unclear as to what exactly @b-mehta wanted here. I've golfed this proof to get it as short as I can make it. Hopefully this will no longer trigger the "wow, that seems long" reaction, although I didn't find the process particularly "fun" :smile:.

I'm not aware of anything in the library that would have helped, so I used `rewrite_search` and some carefully targeted `conv` to write a non-`calc` version. As a result, this is less repairable than my original, but that does not seem to be a priority, and the original still exists in the commit history of this PR in case it does break in the future.

I'm going to resolve this since most of it is off-topic, and it might discourage other reviewers. Feel free to uncollapse.

b-mehta replied:
I think @j-loreaux clarified precisely what I meant very well, in particular that
"a better understanding of what the key steps / transformations are, and which are not so important. Sometimes you even find that steps you thought were important actually weren't". 
My intention with this comment was not to criticise you, nor your proof, and I am sorry that it came across that way! Instead, it was just an immediate reaction that I had somewhat expected a more direct proof for this statement: simpler in the sense of conceptual ideas in the proof rather than proof lines. As you correctly pointed out, term-mode proofs can be both harder to read and harder to maintain, but if there are many ideas in one proof then it can be desirable to take some of them out into their own statements: there's a good chance that a useful idea here might be useful to someone else. And this would of course result in a shorter proof for your lemma as well.
But overall, the new proof you have here does look very clean to me, so thanks for your efforts!

ecstatic-morse committed:
remove `sum_Ico_empty` and `sum_Ico_single`

ecstatic-morse committed:
Golf proof and overhaul names

ecstatic-morse committed:
Add `prod_Ico_add'` with consistent ordering

The old `prod_Ico_add` is kept for API compatibility, but I've expressed
it with a single `@[to_additive]` lemma.

ecstatic-morse committed:
Move `sum_Ico_sub` up, rename, express with `to_additive`

ecstatic-morse committed:
rewrite with only `sum_Ico_add'`

ecstatic-morse committed:
remove unnecessary suffix

j-loreaux commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +169,67 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+  begin
+    conv in n { rw ←nat.sub_add_cancel (nat.one_le_of_lt hmn) },
+    rw ←sum_Ico_add',
+  end,
```
In mathlib, after a `have`, if you want tactic mode just use braces to focus the goal.
```suggestion
  have h₁ : ∑ i in Ico (m+1) n, (f i * G i) = ∑ i in Ico m (n-1), (f (i+1) * G (i+1)),
  { conv in n { rw ←nat.sub_add_cancel (nat.one_le_of_lt hmn) },
    rw ←sum_Ico_add' },
```

j-loreaux commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +169,67 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+    conv in n { rw ←nat.sub_add_cancel (nat.one_le_of_lt hmn) },
+    rw ←sum_Ico_add',
+  end,
+
```
I don't think we generally include blank lines within proofs. The style guidelines ensure it is still readable without them. Should be fixed in two places.
```suggestion
```

j-loreaux commented on src/algebra/big_operators/intervals.lean:
```diff
@@ -164,4 +169,67 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
+  ∑ i in Ico m (n+1), f i - f n = ∑ i in Ico m n, f i :=
+sub_eq_iff_eq_add.mpr $ sum_Ico_succ_top hmn _
+
+end group
```
The multiplicative versions with `to_additive`
```suggestion
section group

-- The partial products (starting from `0`) of `f` and `g` respectively.
local notation `F` n:80 := ∏ i in range n, f i
local notation `G` n:80 := ∏ i in range n, g i

variable [comm_group β]

@[to_additive]
lemma prod_range_succ_div_prod : F (n+1) / F n = f n :=
div_eq_iff_eq_mul'.mpr $ prod_range_succ f n

@[to_additive]
lemma prod_range_succ_div_top : F (n+1) / f n = F n :=
div_eq_iff_eq_mul.mpr $ prod_range_succ f n

@[to_additive]
lemma prod_Ico_div_bot (hmn : m < n) : (∏ i in Ico m n, f i) / f m = ∏ i in Ico (m+1) n, f i :=
div_eq_iff_eq_mul'.mpr $ prod_eq_prod_Ico_succ_bot hmn _

@[to_additive]
lemma prod_Ico_succ_div_top (hmn : m ≤ n) :
  (∏ i in Ico m (n+1), f i) / f n = ∏ i in Ico m n, f i :=
div_eq_iff_eq_mul.mpr $ prod_Ico_succ_top hmn _

end group

-- The partial sums (starting from `0`) of `f` and `g` respectively.
local notation `F` n:80 := ∑ i in range n, f i
local notation `G` n:80 := ∑ i in range n, g i
```

j-loreaux reviewed (commented):
This looks a lot more like mathlib style than the first version. Subject to the changes suggested below I think it looks pretty good.

ecstatic-morse committed:
Apply suggestions from code review

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

ecstatic-morse committed:
final cleanup
